internal/ppa/ppa.go

```diff
@@ -0,0 +1,377 @@
+	basePPA := BasePPA{
+		PPA:    imageDefPPA,
+		Series: series,
+	}
+
+	if deb822 {
+		return &PPA{
+			PPAPrivateInterface: &Deb822PPA{
+				BasePPA: basePPA,
+			},
+		}
+	}
+
+	return &PPA{
+		PPAPrivateInterface: &LegacyPPA{
+			BasePPA: basePPA,
+		},
+	}
+}
+
+// BasePPA holds fields and methods common to every PPAPrivateInterface implementation
+type BasePPA struct {
+	*imagedefinition.PPA
+	Series     string
+	signingKey string
+}
+
+func (p *BasePPA) FullName() string {
+	return p.Name
+}
+
+func (p *BasePPA) name() string {
+	return strings.Split(p.Name, "/")[1]
+}
+
+func (p *BasePPA) user() string {
+	return strings.Split(p.Name, "/")[0]
+}
+
+func (p *BasePPA) url() string {
+	var baseURL string
+	if p.Auth == "" {
+		baseURL = "https://ppa.launchpadcontent.net"
+	} else {
+		baseURL = fmt.Sprintf("https://%s@private-ppa.launchpadcontent.net", p.Auth)
+	}
+	return fmt.Sprintf("%s/%s/%s/ubuntu", baseURL, p.user(), p.name())
+}
+
+// removePPAFile removes the PPA file from the sources.list.d directory
+func (p *BasePPA) removePPAFile(basePath string, fileName string) error {
+	sourcesListD := filepath.Join(basePath, sourcesListDPath)
+	if p.KeepEnabled == nil {
+		return imagedefinition.ErrKeepEnabledNil
+	}
+
+	if *p.KeepEnabled {
+		return nil
+	}
+
+	ppaFile := filepath.Join(sourcesListD, fileName)
+	err := osRemove(ppaFile)
+	if err != nil {
+		return fmt.Errorf("Error removing %s: %s", ppaFile, err.Error())
+	}
+	return nil
+}
+
+func (p *BasePPA) importKey(basePath string, ppaFileName string, ascii bool, debug bool) (err error) {
+	trustedGPGD := filepath.Join(basePath, trustedGPGDPath)
+	keyFileName := strings.Replace(ppaFileName, ".list", ".gpg", 1)
+	keyFilePath := filepath.Join(trustedGPGD, keyFileName)
+
+	err = p.ensureFingerprint(lpBaseURL)
+	if err != nil {
+		return err
+	}
+
+	tmpGPGDir, err := p.createTmpGPGDir(basePath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		tmpErr := osRemoveAll(tmpGPGDir)
+		if tmpErr != nil {
+			if err != nil {
+				err = fmt.Errorf("%s after previous error: %w", tmpErr.Error(), err)
+			} else {
+				err = fmt.Errorf("Error removing temporary gpg directory \"%s\": %s", tmpGPGDir, tmpErr.Error())
+			}
+		}
+	}()
+
+	tmpASCIIKeyFilName := keyFileName + ".asc"
+	tmpASCIIKeyFilePath := filepath.Join(tmpGPGDir, tmpASCIIKeyFilName)
+
+	commonGPGArgs := []string{
+		"--no-default-keyring",
+		"--no-options",
+		"--batch",
+		"--homedir",
+		tmpGPGDir,
+		"--secret-keyring",
+		filepath.Join(tmpGPGDir, "tempring.gpg"),
+		"--keyserver",
+		"hkp://keyserver.ubuntu.com:80",
+	}
+	recvKeyArgs := append(commonGPGArgs, "--recv-keys", p.Fingerprint)
+
+	exportKeyArgs := make([]string, 0)
+	exportKeyArgs = append(exportKeyArgs, commonGPGArgs...)
+
+	if ascii {
+		exportKeyArgs = append(exportKeyArgs, "-a", "--output", tmpASCIIKeyFilePath)
+	} else {
+		exportKeyArgs = append(exportKeyArgs, "--output", keyFilePath)
```

Comment:
Essentially instead of exporting we would then dearmor the key we received. It's preferable to avoid relying on a `gpg` being present inside the image because it might not be, but I can't tell if this uses it inside or on the host building it (then you can just depend on it)

Reply:
You can also dearmor the key yourself, APT uses this awk script:

```
    awk '{ gsub(/\r/,"") }
/^-----BEGIN/{ x = 1; }
/^$/{ if (x == 1) { x = 2; }; }
/^[^=-]/{ if (x == 2) { print $0; }; }
/^-----END/{ x = 0; }' | base64 -d
```

Essentially you can see you just skip the BEGIN and END lines, and then parse every non-empty line not starting with = or - between it and then base64 decode it, it's quite trivial and more reliable than having to spawn gpg.

Reply:
Note that bionic and newer do support `.asc` files in trusted.gpg.d